File: JavaPractice/src/MeetingRooms.java
/**
 * @author divyesh.shah
 */

import java.util.*;

/*
Given N rooms and a meeting list, return the number of the room that held the most meetings.
- Rooms are numbered from 0 to N - 1.
- A meeting in the list is given as a pair of [start_time, end_time) : All the values of start_time are integers and unique.
- The unused room with the lowest number will be assigned to each meeting.

Ex 1>
   Input: N = 2   meetings = [[0, 17], [1, 5], [5, 7]]

1     0: [0, 17]
2     1: [5, 7]

   Output: 1

   Explanation:
   Time 0: Room 0 is assigned to the first meeting.
   Time 1: Room 1 is assigned to the second meeting.
   Time 5: Room 1 is assigned to the third meeting because room 0 is unavailable.

   Room 0 is assigned once, and room 2 is assigned twice. So the output is 1.

Ex 2>
    Input: N = 2 , meetings = [[0,6],[1,5],[7,8]]
    Output: 0

u
2   0: [7, 8]
1   1:

Ex 3>
   Input: N=4  meetings = [[0,12],[1,5],[2,7],[3,4],[4,5]]
   Output: 3

Ex 4>
   Input: N=2 meetings = [[0,1],[1,2],[2,20],[3,4],[4,5],[5,8],[9,13]]
   Output: 1

*/

// package whatever; // don't place package name!
// Installed Libraries: JSON-Simple, JUNit 4, Apache Commons Lang3

class MeetingRooms {
    public static void main (String[] args) {
        System.out.println("Hello Java");
        int N1 = 2;
        int [][] meetings1 = {{0, 17}, {1, 5}, {5, 7}};

        int N2 = 4;
        int [][] meetings2 = {{0,12},{1,5},{2,7},{3,4},{4,5}};

        int N = 2;
        int [][] meetings = {{0, 6}, {1, 5}, {7, 8}};

        System.out.println("Max Rooms " + findMaxRooms(N, meetings));

//        System.out.println("Max Rooms " + findMaxMeetingRooms(N, meetings));
//        System.out.println("Max Rooms " + findMaxMeetingRooms(N1, meetings1));
//        System.out.println("Max Rooms " + findMaxMeetingRooms(N2, meetings2));
//        System.out.println("Max Rooms " + findMaxRooms(N2, meetings2));
//        System.out.println("No Rooms " + findMaxMeeting(N, meetings));
    }

    public static int findMaxRooms(int num, int[][] meetings) {

        if(num == 0 || meetings == null)
            throw new IllegalArgumentException("Pass valid input");

        Arrays.sort(meetings, (a,b) -> a[0] - b[0]);

        PriorityQueue<Integer> roomEndTimes = new PriorityQueue<>();
        Stack<Integer> room = new Stack<>();

        int maxRooms = 0;

        for(int index = 0; index < meetings.length; index++) {
            int startTime = meetings[index][0];
            int endTime = meetings[index][1];
            //room.add(maxRooms);

            if(!roomEndTimes.isEmpty() && roomEndTimes.peek() <= startTime) {
                roomEndTimes.poll();
                room.add(maxRooms--);
            }
            else {
                //maxRooms++;
                room.add(maxRooms++);
            }

            roomEndTimes.offer(endTime);
        }
        return room.peek();

    }

    public static int findMaxMeetingRooms(int num, int[][] meetings) {

        if(num == 0 || meetings == null)
            throw new IllegalArgumentException("Pass valid input");

        Arrays.sort(meetings, (a,b) -> a[0] - b[0]);
        int[] roomCounter = new int[num];
        for(int[] meeting : meetings) {
            int startTime = meeting[0];
            int endTime = meeting[1];

            boolean roomFound = false;

            for(int index = 0; index < num; index++) {

                if(roomCounter[index] <= startTime) {
                    roomCounter[index] = endTime;
                    roomFound = true;
                    break;
                }
            }

            if(!roomFound) {
                roomCounter[num - 1] = endTime;
            }
        }

        int maxMeetings = 0;
        int maxRoom = 0;

        for(int index = 0 ; index < num; index++) {
            if(roomCounter[index] != 0) {
                int meetingCount = index + 1;

                if(meetingCount > maxMeetings) {
                    maxMeetings = meetingCount;
                    maxRoom = index;
                }
            }
        }
        return maxRoom;
    }

    public static int findMaxMeeting(int num, int[][] meetings) {
        if(num == 0 || meetings == null)
            throw new IllegalArgumentException("Pass valid input");

        Arrays.sort(meetings, (a,b) -> a[0] - b[0]);
        for(int i=0; i<meetings.length; i++) {
            for(int j=0; j < meetings[i].length; j++)
                System.out.println(meetings[i][j]);
        }
        return 1;
    }
}

File: PythonPractice/src/MeetingRooms.py
from typing import List
import heapq

def find_max_rooms(num: int, meetings: List[List[int]]) -> int:
    """
    Find the room with the maximum number of meetings.

    Args:
    num (int): Number of rooms.
    meetings (List[List[int]]): List of meetings with [start_time, end_time).

    Returns:
    int: The room number that held the most meetings.
    """
    if num <= 0 or meetings is None:
        raise ValueError("Pass valid input")

    # Sort meetings by start time
    meetings.sort(key=lambda x: x[0])

    # Priority queue to track end times of meetings
    room_end_times = []
    # Stack to track room assignments
    room = []

    max_rooms = 0

    for start_time, end_time in meetings:
        # Free up rooms that have ended before the current meeting starts
        while room_end_times and room_end_times[0] <= start_time:
            heapq.heappop(room_end_times)
            room.pop()

        if len(room) < num:
            room.append(max_rooms)
            max_rooms += 1

        heapq.heappush(room_end_times, end_time)

    return room[-1] if room else -1

def find_max_meeting_rooms(num: int, meetings: List[List[int]]) -> int:
    """
    Find the room with the most meetings by tracking meeting room end times.

    Args:
    num (int): Number of rooms.
    meetings (List[List[int]]): List of meetings with [start_time, end_time).

    Returns:
    int: The room number that held the most meetings.
    """
    if num <= 0 or meetings is None:
        raise ValueError("Pass valid input")

    # Sort meetings by start time
    meetings.sort(key=lambda x: x[0])
    room_counter = [0] * num

    for start_time, end_time in meetings:
        room_found = False

        for index in range(num):
            if room_counter[index] <= start_time:
                room_counter[index] = end_time
                room_found = True
                break

        if not room_found:
            room_counter[num - 1] = end_time

    max_meetings = 0
    max_room = 0

    for index in range(num):
        if room_counter[index] != 0:
            meeting_count = index + 1
            if meeting_count > max_meetings:
                max_meetings = meeting_count
                max_room = index

    return max_room

def find_max_meeting(num: int, meetings: List[List[int]]) -> int:
    """
    Print the meetings and return 1 for demonstration purposes.

    Args:
    num (int): Number of rooms.
    meetings (List[List[int]]): List of meetings with [start_time, end_time).

    Returns:
    int: 1 for demonstration purposes.
    """
    if num <= 0 or meetings is None:
        raise ValueError("Pass valid input")

    for meeting in meetings:
        for time in meeting:
            print(time)

    return 1

# Example usage
if __name__ == "__main__":
    N1 = 2
    meetings1 = [[0, 17], [1, 5], [5, 7]]

    N2 = 4
    meetings2 = [[0,12],[1,5],[2,7],[3,4],[4,5]]

    N = 2
    meetings = [[0, 6], [1, 5], [7, 8]]

    print("Max Rooms:", find_max_rooms(N, meetings))
    print("Max Rooms by Meeting Rooms:", find_max_meeting_rooms(N, meetings))
    print("Max Meeting:", find_max_meeting(N, meetings))
